everviz/plugins/config_editor/config_editor.py

```diff
@@ -0,0 +1,172 @@
+
+        md_content = f"""```
+{config_str}
+```
+"""
+        return [
+            dcc.Textarea(id=self.text_area, value=config_str, hidden=not self.editing,),
+            dcc.Markdown(id=self.md_area, children=md_content,),
+        ]
+
+    def render_buttons(self):
+        if not self.editing:
+            return [
+                html.Button("Edit", id=self.btn_edit),
+                # Had to wrap the button in a div because the hidden option was not
+                # working on the button element
+                html.Div([html.Button("Cancel", id=self.btn_cancel),], hidden=True),
+                html.Div(
+                    dcc.Upload(html.Button("Upload File"), id=self.btn_upload),
+                    style={"display": "inline-block",},
+                ),
+                html.Div(
+                    [html.Button("Reset", id=self.btn_reset),],
+                    hidden=not self.default_conf_path.exists(),
+                ),
+            ]
+        return [
+            html.Button("Save", id=self.btn_edit),
+            html.Button("Cancel", id=self.btn_cancel),
+            html.Div(html.Button("Reset", id=self.btn_reset,), hidden=True),
+            dcc.Upload(children=None, id=self.btn_upload),
+        ]
+
+    @property
+    def layout(self):
+        return html.Div(
+            [
+                html.H1(children=self.title, style={"textAlign": "center"},),
+                html.Div(
+                    [
+                        html.Div(
+                            id=self.config_comp,
+                            children=self.render_text_component(),
+                            style={
+                                "width": "100%",
+                                "height": "100%",
+                                "display": "inline-block",
+                            },
+                        ),
+                        html.Div(
+                            id=self.editor_buttons,
+                            children=self.render_buttons(),
+                            style={"textAlign": "center"},
+                        ),
+                    ]
+                ),
+            ]
+        )
+
+    def set_callbacks(self, app):
+        @app.callback(
+            [
+                Output(self.config_comp, "children"),
+                Output(self.editor_buttons, "children"),
+            ],
+            [
+                Input(self.btn_cancel, "n_clicks"),
+                Input(self.btn_edit, "n_clicks"),
+                Input(self.btn_reset, "n_clicks"),
+                Input(self.btn_upload, "contents"),
+            ],
+            [State(self.text_area, "value")],
+        )
+        def update_layout(
+            cancel_clicks, edit_clicks, reset_clicks, upload_content, value
+        ):
+            changed_id = [p["prop_id"] for p in callback_context.triggered][0]
+            if edit_clicks and self.btn_edit in changed_id:
```

Comment:
Out of curiosity: why is it necessary to both check `edit_clicks` and `changed_id`?

Reply:
this was recommended way to do it in a dash tutorial